### PR TITLE
Migrate integration tests to Barometer tests that should run as a part of every gradle test run

### PR DESCRIPTION
**Pull request by tdaffin**
Finally posting the integration tests we discussed in PR#559.

This includes removal of the CommandTest workaround.

Also, I could not get the Barometer author's maven repo to work, so I set up my own at https://dl.bintray.com/tdaffin/mcmods.  Note to @jjtParadox I think the problem is that gradle expects the repo to be structured with a particular path.  To get mine to work I had to upload to a path of /com/jjtparadox/barometer/barometer/, so I think the /com/jjtparadox/barometer corresponds to the 'group' and the final /barometer to the 'artifact' that maven expects...  Something like that anyway.  The point is that the jars then turn up under https://dl.bintray.com/tdaffin/mcmods/com/jjtparadox/barometer/barometer/0.0.3-1.12/ 

I have not run this in a clean clone using only setupCIWorkspace, but I think that will happen as part of the PR verification?

### CODE REVIEW

**Comment by rubensworks** [comment]
It looks like an exception occurred while testing:
```
java.lang.SecurityException: Cannot replace the FML security manager
	at net.minecraftforge.fml.relauncher.FMLSecurityManager.checkPermission(FMLSecurityManager.java:54)
	at java.lang.System.setSecurityManager0(System.java:300)
	at java.lang.System.setSecurityManager(System.java:291)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:77)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:45)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:83)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:35)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:117)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:62)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:62)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:67)
Exception in thread "main" net.minecraftforge.fml.relauncher.FMLSecurityManager$ExitTrappedException
	at net.minecraftforge.fml.relauncher.FMLSecurityManager.checkPermission(FMLSecurityManager.java:49)
	at java.lang.SecurityManager.checkExit(SecurityManager.java:761)
	at java.lang.Runtime.exit(Runtime.java:107)
	at java.lang.System.exit(System.java:971)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:71)
```

I assume this should have failed the build as well?

File: build.gradle
Comment (rubensworks) [inline build.gradle]: Can't this be made the default dir in barometer, as this will almost always be just this value.

File: build.gradle
Comment (rubensworks) [inline build.gradle]: Should not be necessary here as well.

File: build.gradle
Comment (rubensworks) [inline build.gradle]: Should also be abstractable.

File: build.gradle
Comment (rubensworks) [inline build.gradle]: I would hide this after a config option in barometer, something like `barometer.acceptEula()`.

File: build.gradle
Comment (rubensworks) [inline build.gradle]: Not sure what this does, is it needed?

File: gradle/forge.gradle
Comment (rubensworks) [inline gradle/forge.gradle:49]: What does this do? Can it be removed?

File: gradle/forge.gradle
Comment (rubensworks) [inline gradle/forge.gradle]: It should be possible to automatically include these deps by declaring them in the pom of barometer.

**Review by rubensworks — CHANGES_REQUESTED** [review]
Thanks!

I mainly have some concerns about boilerplate config that should not really be included in this repo, but should be abstractable to barometer itself.

**Comment by jjtParadox** [comment]
Most of the boilerplate is because I wanted to give options to those using Barometer, but you're right in that some of this can be abstracted into Barometer itself. I'll do my best to respond to your comments in the gradle file.

File: build.gradle
Comment (jjtParadox) [reply build.gradle]: The `workingDir` property is a part of Gradle's testing job, not Barometer. If not set, it defaults to the project directory, which causes a bit of a mess when Barometer starts the test server in that place.

I had originally intended for Barometer to only depend on JUnit and not have any direct interaction with Gradle, but I now realize that I already depend on ForgeGradle already, so perhaps having some cleaner integration with Gradle would not be unnecessary bloat.

Figuring out how to properly integrate with Gradle may take some time as I have no experience with it, but I'll do what I can.

File: build.gradle
Comment (jjtParadox) [reply build.gradle]: Yeah this is something Barometer could handle. I just need to figure out how to make Barometer into a Gradle plugin. (Same goes for https://github.com/CyclopsMC/IntegratedDynamics/pull/562#discussion_r175274508 and https://github.com/CyclopsMC/IntegratedDynamics/pull/562#discussion_r175274521)

File: gradle/forge.gradle
Comment (jjtParadox) [reply gradle/forge.gradle:49]: This line is because Barometer has a light dependency on ForgeGradle, as it uses part of ForgeGradle to start the test server.

`makeStart` is the ForgeGradle task that creates the `main()` classes that run the client and server in dev environments (i.e. the intellij run tasks), and Barometer piggy backs off of those classes to properly start a test server (See https://github.com/jjtParadox/Barometer/blob/1.12/src/main/kotlin/com/jjtparadox/barometer/tester/Tests.kt).

If I can figure out how to make Barometer a gradle plugin and an extension of ForgeGradle, this might not be necessary in the future.

File: gradle/forge.gradle
Comment (jjtParadox) [reply gradle/forge.gradle]: That's my fault, as I still lack experience in publishing to maven repos :pensive: I'll try again at getting my maven repo working.

@tdaffin if you figure out how to get the pom working, please let me know! I'd appreciate the help :smile: 

EDIT: Figured it out!

**Comment by jjtParadox** [comment]
To summarize, getting most of this out of `build.gradle` and into Barometer requires turning Barometer into a gradle plugin. It is certainly possible, but it's also a rabbit hole that will probably take a while to navigate. I'll figure out what I can (and I'm still very open to PRs! They help a lot :smile:)

**Comment by rubensworks** [comment]
> To summarize, getting most of this out of build.gradle and into Barometer requires turning Barometer into a gradle plugin.

A Barometer gradle plugin would indeed be great to have!

I would prefer to wait until such a plugin would exist, so that the build file can be kept as clean as possible :)

I'd love to help implementing this, but I'm afraid I already lack sufficient time to just maintain my mods :/ But who knows, perhaps I find a sea of time some day...

**Comment by rubensworks** [comment]
@jjtParadox Side-question: does Barometer work in IDEs like Intellij? If not, is that planned for the future?

**Comment by jjtParadox** [comment]
> does Barometer work in IDEs like Intellij?

It does! You can get the usual JUnit integration by running the `test` task from the gradle window:

image

Cool thing about Barometer and JUnit integration is it also works with Travis and other CI systems, and thus GitHub's checks. Could be helpful for PRs!

**Comment by jjtParadox** [comment]
Also, I think I have Bintray working properly now! https://bintray.com/jjtparadox/MC/Barometer (should make `build.gradle` a _little_ cleaner now)

<blockquote><div><strong><a href="https://bintray.com/jjtparadox/MC/Barometer">Barometer</a></strong></div><div>Download Barometer from Bintray. Modern Software Distribution Service for Docker, Maven, Debian, RPM, npm, NuGet and more.</div></blockquote>

**Comment by jjtParadox** [comment]
Oh yeah and the `java.lang.SecurityException: Cannot replace the FML security manager` thing is because Forge gets fussy that the Gradle testing environment is managing how it runs and exits. It's completely harmless and never impacts actual testing, but I haven't yet figured out how to get rid of the message...

**Comment by rubensworks** [comment]
@jjtParadox Awesome :)

> Oh yeah and the java.lang.SecurityException: Cannot replace the FML security manager thing is because Forge gets fussy...

Perhaps you could have a look at the Forge CI. They also run a minimal instance for testing I think. I think they don't get such an error in the logs.

**Comment by jjtParadox** [comment]
Oh wow didn't even know Forge did that.

After doing some skimming, it looks like they don't start up an entire server instance and instead just initialize all the relevant MC classes so it can pretend to be running in a loaded environment (which is closer to how unit testing actually should work). One limitation of this is it appears Forge's CI cannot tick a world during testing.

I'll look into if Barometer can learn anything from Forge's CI, as there are quite a few handy things in there I could use.

File: gradle/forge.gradle
Comment (tdaffin) [reply gradle/forge.gradle]: I pushed the change to use @jjtParadox 's repository, along with the removal of these now unnecessary lines.

**Comment by jjtParadox** [comment]
@rubensworks what do you usually use tests for? Considering that most of the IntegratedDynamics tests do not require a full ticking world, it may be more appropriate to use Forge's TestRunner (which can be found in src/test in the Forge repo) for these instead of Barometer.

File: build.gradle
Comment (tdaffin) [reply build.gradle]: This causes the test task to always run, even if there have been no code changes since the last run.
I find this useful when debugging tests as I sometimes want to run them multiple times without making code changes.
However, I'll go ahead and remove it as it isn't really necessary for the CI pipeline.

File: gradle/forge.gradle
Comment (jjtParadox) [inline gradle/forge.gradle]: This likely needs to be `com.jjtparadox.barometer:Barometer:0.0.3-1.12` (Note the capital B)

Comment (tdaffin) [reply gradle/forge.gradle]: Oops... I think it was picking up the lowercase version from my mavenLocal repository...

**Comment by rubensworks** [comment]
@jjtParadox For the current tests, world ticking is not required.
I don't see any future tests where world ticking would be required.

I thought Barometer was a standalone version of the [ForgeTestRunner](https://github.com/MinecraftForge/MinecraftForge/blob/16bfd8cef1d12ee9ca0de1122addaf9916767ae9/src/test/java/net/minecraftforge/fml/common/registry/ForgeTestRunner.java), with some additions, such as world ticking.

In any case, I don't want to copy paste that class in this mod for running tests, as I would like to unit tests all my other mods more extensively as well, and I don't like duplicate code to exist between my mods :)

If you think Barometer is not the appropriate platform, then perhaps we/I should look into creating a standalone variant of ForgeTestRunner.

**Comment by jjtParadox** [comment]
You raise a good point about code reuse. At first I had thought that ForgeTestRunner would have been available from Forge for authors to use, but test classes are not usually compiled into libraries.

Barometer does start up an entire server instance to test in, but if you have no issue with the test execution time then we can certainly continue using it :smile:

**Comment by jjtParadox** [comment]
Well, I think I got a Gradle plugin working. Check version 0.0.4.

Basically, do this in your gradle file: 

```gradle
// Top of file
buildscript {
    repositories {
        // ...
        // Forge's stuff here...
        // ...
        maven { url "https://dl.bintray.com/jjtparadox/MC" }
    }
    dependencies {
        // More Forge stuff
        // ...
        classpath 'com.jjtparadox.barometer:Barometer:0.0.4-1.12'
    }
}

apply plugin: 'net.minecraftforge.gradle.forge'
apply plugin: 'com.jjtparadox.barometer' // Must be after the forge plugin

dependencies {
    // Other dependencies
    // ...
   testCompile 'com.jjtparadox.barometer:Barometer:0.0.4-1.12'
}

barometer {
    acceptEula()
    testServerDir = '/testing' // Optional
}
        
```

Then running `gradle test` should run all the tests under Barometer. (I'm looking into if I can remove that extra `testCompile` dependency)

Disclaimer: I have not been able to test this yet. Please let me know ~~if~~ when it explodes.

**Comment by tdaffin** [comment]
I've been testing 0.0.4 on my 'ExampleMod' and I've found that I also need to add
```
repositories {
    maven { url "https://dl.bintray.com/jjtparadox/MC" }
}
```
After `apply plugin: 'com.jjtparadox.barometer'` and before `dependencies {`
However, even with that I get these errors when trying to run `./gradlew test`:
```
examplemod.ExampleModTest > initializationError FAILED
    java.lang.IncompatibleClassChangeError

examplemod.TestUtilsTest > classMethod FAILED
    kotlin.UninitializedPropertyAccessException

net.nuffin.not.DummyTest > classMethod FAILED
    kotlin.UninitializedPropertyAccessException

net.nuffin.mc.BuilderTest > classMethod FAILED
    kotlin.UninitializedPropertyAccessException
```

I'm going to try again with the code in the PR and see if I get the same problems.

**Comment by tdaffin** [comment]
I tried running `./gradlew test` on the v0.0.4-1.12 tag of barometer and got some similar errors:
```

com.jjtparadox.barometer.test.BarometerExampleTest > initializationError FAILED
    java.lang.IncompatibleClassChangeError

com.jjtparadox.barometer.test.BarometerExampleTestJava > classMethod FAILED
    kotlin.UninitializedPropertyAccessException

com.jjtparadox.barometer.otherpackage.test.BarometerExampleTest > classMethod FAILED
    kotlin.UninitializedPropertyAccessException
```
Perhaps I should open an issue on Barometer?

**Comment by jjtParadox** [comment]
Yeah do that, and make sure to include the --stacktrace flag when you run the command. This looks similar to an issue I had previously, but I thought I fixed it...

**Comment by tdaffin** [comment]
@jjtParadox  ok, issue submitted: https://github.com/jjtParadox/Barometer/issues/11

**Comment by jjtParadox** [comment]
Fixed with v0.0.6

**Comment by tdaffin** [comment]
Ok, 0.0.6 seems to work on my ExampleMod.
One thing though -- I also had to add
```
repositories {
    maven {
        url "https://dl.bintray.com/jjtparadox/MC"
    }
}
```
After the apply plugin and before the dependencies.
I'll post a PR to Barometer to add that to the README in a minute.
Also, I'll work on updating this branch to 0.0.6 after that.

**Comment by tdaffin** [comment]
Ok, done: https://github.com/jjtParadox/Barometer/pull/13

**Comment by rubensworks** [comment]
I have the feeling this PR is starting to look too much like a chat room 😄

Perhaps it's better to move this development discussion to a different place? Feel free to use the #development channel in the [Cyclops Discord server](https://discord.gg/9yDxubB) for that.

**Comment by tdaffin** [comment]
All the requested changes have now been made -- although it looks like the security exception is still there:

```
java.lang.SecurityException: Cannot replace the FML security manager
	at net.minecraftforge.fml.relauncher.FMLSecurityManager.checkPermission(FMLSecurityManager.java:54)
	at java.lang.System.setSecurityManager0(System.java:300)
	at java.lang.System.setSecurityManager(System.java:291)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:77)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:45)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:83)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:35)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:117)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:62)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:62)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:67)
Exception in thread "main" net.minecraftforge.fml.relauncher.FMLSecurityManager$ExitTrappedException
	at net.minecraftforge.fml.relauncher.FMLSecurityManager.checkPermission(FMLSecurityManager.java:49)
	at java.lang.SecurityManager.checkExit(SecurityManager.java:761)
	at java.lang.Runtime.exit(Runtime.java:107)
	at java.lang.System.exit(System.java:971)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:71)
```

**Comment by jjtParadox** [comment]
We won't stray too far from the topic of this PR, we promise 😜

I believe that now with Barometer having its own gradle plugin, most the issues raised in the review are addressed. I still haven't found anything to do about the security exception short of ASM modifications, but the exception is harmless. If you feel that is a barrier to merging this PR I can look into fixing it.

File: src/test/java/org/cyclops/integrateddynamics/core/evaluate/variable/integration/TestBlockOperators.java
Comment (rubensworks) [inline src/test/java/org/cyclops/integrateddynamics/core/evaluate/variable/integration/TestBlockOperators.java]: I suggest hardcoding the sound event names, so that this test can also run client-side.

**Comment by rubensworks** [comment]
> If you feel that is a barrier to merging this PR I can look into fixing it.

I'd rather not have any exceptions shown in builds... (That reminds me to find out why `extractAnnotationsJarError` throws and exception)

Perhaps it can be solved by configuring your own security manager? You don't really need `FMLSecurityManager` for testing I guess.

**Comment by jjtParadox** [comment]
Done! Version 0.0.7 no longer displays the exceptions.

**Comment by tdaffin** [comment]
Hmm, still seeing the security exception in the CI build here -- even though it did not appear when I ran it locally -- not for the local build and test of Barometer, my ExampleMod or IntegratedDynamics.

@jjtParadox  Perhaps I should setup a test mod here on GitHub and configure a CI pipeline for it in order to test Barometer in a real CI environment.

Exception details:
```
:testClasses
:testPicked up _JAVA_OPTIONS: -Xmx2048m -Xms512m
java.lang.SecurityException: Cannot replace the FML security manager
	at net.minecraftforge.fml.relauncher.FMLSecurityManager.checkPermission(FMLSecurityManager.java:54)
	at java.lang.System.setSecurityManager0(System.java:300)
	at java.lang.System.setSecurityManager(System.java:291)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:77)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:45)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:83)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:35)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:117)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:62)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:62)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:67)
Exception in thread "main" net.minecraftforge.fml.relauncher.FMLSecurityManager$ExitTrappedException
	at net.minecraftforge.fml.relauncher.FMLSecurityManager.checkPermission(FMLSecurityManager.java:49)
	at java.lang.SecurityManager.checkExit(SecurityManager.java:761)
	at java.lang.Runtime.exit(Runtime.java:107)
	at java.lang.System.exit(System.java:971)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:71)
```

**Comment by jjtParadox** [comment]
...what on earth?

@rubensworks if you run this PR locally, do you get the exception too?
@tdaffin don't worry about setting up a test CI, I can take care of that (and I don't want to turn this PR into a chatroom at ruben's request :stuck_out_tongue:)

**Comment by jjtParadox** [comment]
...what on earth?

@rubensworks if you run this PR locally, do you get the exception too?
@tdaffin don't worry about setting up a test CI, I can take care of that (and I don't want to turn this PR into a chatroom at ruben's request :stuck_out_tongue:)

**Comment by milandamen** [comment]
Sorry for disturbing this PR, but I came across the FML security manager exception as well. It seems to be fixed in a newer version of Gradle (though I don't know whether it's possible to even use a newer version of gradle).

See:
https://github.com/gradle/gradle/issues/5305
https://github.com/gradle/gradle/pull/5324

Even though this exception gets shown, the exit code of `./gradlew test` is correct. It is 1 when tests fail and 0 when all tests succeed, regardless of whether or not the security exception gets throw. So the exception can basically be ignored.

**Comment by tdaffin** [comment]
Looks like the referenced PR was merged into gradle 4.9, and forgegradle switched to using that in version 3.0.
Checking the build.gradle for forge indicates that they switched to forgegradle 3.x in branch 1.13, so this PR might be good to go when ID is updated to 1.13 or later.

**Comment by rubensworks** [comment]
FYI, I plan to look into updating to 1.13 in a couple of weeks. But it's going to be a big update, so I don't expect this to be done any time soon though.

**Comment by rubensworks** [comment]
FYI, the gradle issue should be fixed now in the [`master-1.15`](https://github.com/CyclopsMC/IntegratedDynamics/tree/master-1.15) branch.

**Comment by tdaffin** [comment]
It has been a while since I've done any Java development, but I'll see if I can get this to work again.

**Comment by tdaffin** [comment]
I figured I would merge the latest 1.12 branch into this branch to get back up to speed on things, and strangely enough that seems to have been enough to get the tests to all run correctly!
Perhaps the gradle fix was ported back to the 1.12 branch of forge at some point?
I'm still up for trying to port these changes to 1.15, but I'm also happy to have this PR merged to 1.12 if you think it is ready now.

**Comment by rubensworks** [comment]
Thanks for looking into this @tdaffin!
I just checked out your branch, and everything appears to be working nicely.

I just have two more concerns before merging this:

1. Intellij's test runner (when running specific tests, not via gradle) doesn't appear to be working correctly with this. I see that the integration tests are starting up a Forge server, but the EULA doesn't appear to be created, so it crashes. This type of workflow is quite essential for my development, so it would be very unfortunate if we lost that.
2. Now that I'm porting ID to 1.15, how does Barometer work on 1.15? If it doesn't work there, these changes would have to be reverted again. If it is independent of the MC version, this would be perfect of course.

**Comment by tdaffin** [comment]
I took a look at IntelliJ's problems this morning.  For me, the tests appear to complete but the GUI hangs on a random test, so that is obviously not ok.
I'm guessing that perhaps something odd in forge/gradle is causing that, but that is a pure shot in the dark.

I've taken some time this morning to create a new branch in my fork that merges the latest 1.15 into this branch:
https://github.com/CyclopsMC/IntegratedDynamics/compare/master-1.15...tdaffin:barometer_tests-1.15

This has led to some problems as well -- in 1.15 I cannot run the tests at all.  It seems that the problem may be due to barometer not being recognized as a gradle plugin in the new version.  I wonder if that is due to 1.15 using a more recent version of grade?  Perhaps @jjtParadox would be interested to know that?

FYI: All the problems are illustrated by the commented out lines in build.gradle, that I put into a separate commit to highlight them: https://github.com/CyclopsMC/IntegratedDynamics/commit/44e28ba2a796433409b58bec1127f1247df0d9c8

**Comment by milandamen** [comment]
If I remember correctly from a few years ago, the IntelliJ tests GUI hangs sometimes when a test fails. You have to run each test separately to find out which one failed though.

**Comment by rubensworks** [comment]
Since this PR has gone stale, I'll close it.
Feel free to re-open it should anyone be willing to continue work on it.

**Pull request closed without merging**
